java/bundles/org.eclipse.set.basis/src/org/eclipse/set/basis/graph/TopPath.java

```diff
@@ -72,49 +113,76 @@ public BigDecimal firstEdgeLength() {
 					.getWert();
 
-			// If the point is on the current edge, check which direction to add
 			if (point.edge().equals(edge)) {
-				// In top direction
-				if (previousEdge == null
-						|| previousEdge.getIDTOPKnotenB().getValue() == edge
-								.getIDTOPKnotenA().getValue()
-						|| previousEdge.getIDTOPKnotenA().getValue() == edge
-								.getIDTOPKnotenA().getValue()) {
-					final BigDecimal pointDistance = pathOffset
-							.add(point.distance());
-					if (pointDistance.compareTo(BigDecimal.ZERO) < 0
-							|| pointDistance.compareTo(length) > 0) {
-						return Optional.empty();
-					}
-
-					return Optional.of(pointDistance);
-				}
-
-				// against top direction
-				final BigDecimal pointDistance = pathOffset.add(edgeLength)
-						.subtract(point.distance());
-				if (pointDistance.compareTo(BigDecimal.ZERO) < 0
-						|| pointDistance.compareTo(length) > 0) {
-					return Optional.empty();
-				}
-
-				return Optional.of(pointDistance);
+				return getDistance(point, pathOffset, previousEdge, edge);
 			}
 
 			// Point not on this edge, check next edge
 			if (previousEdge != null) {
 				pathOffset = pathOffset.add(edgeLength);
 			}
 			previousEdge = edge;
-
 		}
 		return Optional.empty();
 	}
+
+	// If the point is on the current edge, check which direction to add
+	private Optional<BigDecimal> getDistance(final TopPoint point,
+			final BigDecimal pathOffset, final TOP_Kante previousEdge,
+			final TOP_Kante edge) {
+		final BigDecimal edgeLength = edge.getTOPKanteAllg().getTOPLaenge()
+				.getWert();
+		// Point in first Edge
+		if (previousEdge == null) {
+			return pointOnFirstEdge(point, edgeLength);
+		}
+		// In top direction
+		if (previousEdge.getIDTOPKnotenB().getValue() == edge.getIDTOPKnotenA()
+				.getValue()
+				|| previousEdge.getIDTOPKnotenA().getValue() == edge
+						.getIDTOPKnotenA().getValue()) {
+
+			final BigDecimal pointDistance = pathOffset.add(point.distance());
+			if (pointDistance.compareTo(BigDecimal.ZERO) < 0
+					|| pointDistance.compareTo(length) > 0) {
+				return Optional.empty();
+			}
+
+			return Optional.of(pointDistance);
+		}
+
+		// against top direction
+		final BigDecimal pointDistance = pathOffset.add(edgeLength)
+				.subtract(point.distance());
+		if (pointDistance.compareTo(BigDecimal.ZERO) < 0
+				|| pointDistance.compareTo(length) > 0) {
+			return Optional.empty();
+		}
+
+		return Optional.of(pointDistance);
+	}
+
+	private Optional<BigDecimal> pointOnFirstEdge(final TopPoint point,
+			final BigDecimal edgeLength) {
+		if (point.distance().compareTo(startNode.distance()) == 0) {
+			return Optional.of(BigDecimal.ZERO);
+		}
+
+		final boolean inTopDirection = edgeLength.subtract(startNode.distance())
+				.compareTo(firstEdgeLength) == 0;
+		// Point lie before path start
+		if (inTopDirection == point.distance()
+				.compareTo(startNode.distance()) < 0) {
+			return Optional.empty();
+		}
```

Comment:
@TruongQuangSB The equal 0 case makes me struggle a bit when not in top direction. I think we should simply do

```suggestion
		if (
		        (inTopDirection && point.distance().compareTo(startNode.distance()) < 0) ||
		        (!inTopDirection && point.distance().compareTo(startNode.distance()) > 0)
		) {
			return Optional.empty();
		}
```

Reply:
@TruongQuangSB Ahh yes you are right. Have overseen this :smile: